src/main/scala/tech/cryptonomic/conseil/tezos/michelson/contracts/TokenLedgers.scala

```diff
@@ -0,0 +1,185 @@
+package tech.cryptonomic.conseil.tezos.michelson.contracts
+
+import tech.cryptonomic.conseil.tezos.TezosTypes.{AccountId, Contract, ContractId, Decimal, ScriptId}
+import tech.cryptonomic.conseil.tezos.TezosTypes.Micheline
+import tech.cryptonomic.conseil.tezos.michelson.dto.{
+  MichelsonBytesConstant,
+  MichelsonInstruction,
+  MichelsonIntConstant,
+  MichelsonSingleInstruction
+}
+import cats.implicits._
+import scala.util.Try
+import com.typesafe.scalalogging.LazyLogging
+
+/** Collects custom token contracts structures and operations */
+object TokenLedgers extends LazyLogging {
+
+  /** a key hash paired with a balance */
+  type BalanceUpdate = (ScriptId, BigInt)
+
+  /** typed wrapper to clarify the meaning of the numerical id */
+  case class BigMapId(id: BigDecimal) extends AnyVal
+
+  /** alias to the custom code that will read the new balance for a token from big map diffs */
+  private type BalanceReader = Contract.BigMapUpdate => Option[BalanceUpdate]
+
+  /* data structure wrapping useful information + functions to act on ledgers
+   *
+   * mapId is a reference to the big map used to store data on balances
+   * balanceReader is a function that will use a big map update and extract
+   *   balance information from it
+   */
+  private case class LedgerToolbox(
+      mapId: BigMapId,
+      balanceReader: BalanceReader
+  )
+
+  /** For each specific contract available we store a few
+    * relevant bits of data useful to extract information
+    * related to that specific contract shape
+    *
+    * We're assuming here that the risk of collision on contract addresses representing
+    * tokens, even on different tezos networks, is actually zero.
+    */
+  private val ledgers: Map[ContractId, LedgerToolbox] = Map(
+    ContractId("KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux") -> LedgerToolbox(
+          BigMapId(1718),
+          KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux_Balance_Read
+        )
+  )
```

Comment:
This address should not be hard-coded. The list of all known contracts should be read from the registered tokens CSV file.

Reply:
Different Tezos networks will have token contracts at different addresses. These contracts will be hard-coded in the network-specific CSV files.

Reply:
I'm short on ideas here.
If we read dynamically the contract id (and the map id, since we're there), I would anyway need a way to identify the specifics of how to extract the balance for that specific token contract.
This is now written as a custom code here, using the balance-reader function.

I'm not sure how could I dynamically read the balance extraction code too from config.
The alternatives here might be: 
 * use dynamic class loading, where the scala class name is written in the csv file or some other extra configuration
 * use a key (maybe a readable string) and hard-code such key to trace the specific code implementation.

The second option is sort-of in-between, the code is not dynamically loaded, but it can be referred-to dynamically, and thus point to different contracts and big maps on different chains.

Only the first option allows to choose dynamically (at startup time) which code needs to be executed to actually read the balance specifically for each contract on the chain.

Reply:
Perhaps my earlier comment wasn't clear. You have hard-coded in the Scala code the address `KT1RmDuQ6LaTFfLrVtKNcBJkMgvnopEATJux` as a token contract address. Instead, you should be reading these token contract addresses from the CSV file currently located in `src/main/resources/csv/mainnet.csv`. 

In the future, instead of hard-coding these addresses into a CSV file we will read them dynamically from the blockchain.

Reply:
I think your comment was clear, but we might want to talk face-to-face so that I can better explain what're the consequences I see from reading addresses from config or from chain.

Reply:
I think I understand your question better now and have left a comment in our chat.

Reply:
rewritten to read from config